### DISCUSSION TITLE
encoding / decoding JS getMesh() format

### DISCUSSION THREAD

**Discussion by stewartoallen** [elalish/manifold#255] [General]
Is it assumed or required that the mesh is indexed? The example uses ThreeJS indexed geometries. Most of the meshes I deal with come from STLs and are not indexed. It's not hard to index them, but would be nice if it could be avoided.

**Comment by stewartoallen** [comment]
too soon. reading [here](https://manifoldcad.org/docs/html/group___connections.html#structmanifold_1_1_mesh) seems to answer my question. I'll try to spend more time with the library first. lots of questions about performance relating to marshaling between JS and WASM. for the STL case, it would likely be faster to do the indexing in WASM than JS.

**Comment by stewartoallen** [comment]
going in and out of ThreeJS BufferAttribute formats presents a significant amount of time spent in JS (for loops + accessors). would you consider adding a helper method that returns the flattened vertex and index arrays from meshes?

**Reply by pca006132** [reply]
I think the way to go for this is to pass an array buffer directly as vector for js -> wasm part. For wasm -> js, I think we can also do something similar by copying the data from a vector to an array buffer in C++.

**Reply by stewartoallen** [reply]
yes, it would be ideal if JS could provide a properly sized ArrayBuffer to be filled by the library. do you have any reason to think this wouldn't work with SharedArrayBuffers?

**Reply by pca006132** [reply]
No? I think this should work, but I don't yet have time to implement that for now. Maybe you can try to implement it and see how much performance improvement this can get and submit a PR.

**Reply by stewartoallen** [reply]
Would love to. Looking now. It's been > 20 years since I last did anything with C++ :/  When I get this working, it will almost certainly be the "wrong way" :)

**Reply by stewartoallen** [reply]
this is how far this hack has progressed. it definitely works in terms of returning to JS-land a pointer to the data in the wasm heap that contains the written i32 values. and they can be later freed with another call using same pointer. only, the values just contain `p` ("p +" was for testing) and not the values of x, y, and z.

```
int Manifold::GetMeshArray() const {
    const Impl& impl = *GetCsgLeafNode().GetImpl();

    int n = NumVert();
    int *x = (int *)malloc(sizeof(int) * n * 3);

    int p = 0;
    for (auto i = impl.vertPos_.begin(); i != impl.vertPos_.end(); i++) {
        const glm::vec3 v = *i;
        *(x + p) = p + (v.x) * 100000; p++;
        *(x + p) = p + (v.y) * 100000; p++;
        *(x + p) = p + (v.z) * 100000; p++;
    }

    return (int)x;
}
```

**Reply by stewartoallen** [reply]
user error on the JS side :/ so now that I can directly access the vertex/index data from the WASM shared memory, the fun begins!

**Reply by stewartoallen** [reply]
well, direct memory access is more than 110x faster. so that's something. @pca006132 @hrgdavor 

```
let c = Module.sphere(10, 100);
let cm = c.getMesh();
let md = c.getMeshDirect();
console.log({ c, cm, md });
console.time('getMesh');
for (let i=0; i<1000; i++) c.getMesh();
console.timeEnd('getMesh');
console.time(`getMeshDirect`);
for (let i=0; i<1000; i++) c.getMeshDirect();
console.timeEnd(`getMeshDirect`);
```

browser console output:

```
getMesh: 5484.587890625 ms
getMeshDirect: 48.624267578125 ms
```

getMeshDirect() returns flat vertex and index arrays suitable for ThreeJS

**Reply by pca006132** [reply]
can you send a PR for this? would love to review this

**Reply by stewartoallen** [reply]
for sure. it's a bit ugly. and I don't exactly know how to do "proper" memory management. using malloc/free, for example. incoming soon.

**Reply by stewartoallen** [reply]
works for the WASM build, but mem pointers break the normal build. noted in the PR conversation.

**Reply by hrgdavor** [reply]
> well, direct memory access is more than 110x faster. so that's something. @pca006132 @hrgdavor 
> ...
> getMeshDirect() returns flat vertex and index arrays suitable for ThreeJS

I am working on supporting 2 major 3d js engines Babylon and Threejs for jscad (on top of jscad's current support for regl) and there are some nuances and differences between them that also reflect data layout (I think not for vertex data, but for color info).
Looking how resposnive ppl here are, I am sure if needed we could get some options for better supporting Babylon too.

**Reply by stewartoallen** [reply]
this is just dealing with the vertex data, not colors or normals. and index array if the mesh is indexed. do you have details on how the other engines store vertex data, indexed and un-indexed?

**Reply by elalish** [reply]
Every renderer I know of uses indexed verts because it requires so much less GPU memory. But more importantly, the verts must be indexed to represent a manifold mesh - this is the problem with STL (an un-indexed file format). Then the topology (exact) must be derived from the geometry (inexact), which is why an STL cannot reliably transmit manifoldness information.

**Reply by hrgdavor** [reply]
> this is just dealing with the vertex data, not colors or normals. and index array if the mesh is indexed. do you have details on how the other engines store vertex data, indexed and un-indexed?

Babylon and Three use same layout for vertices, but they use oposite CCW-CW so in my prototypes i need to invert meshes when switching between them.

color
 - babylon only uses 4 channel for color data
 - threejs can use RGB and RGBA

I tested this stuff here: https://github.com/jscad/OpenJSCAD.org/discussions/944 and last update on that is a bit old now, as I easily get side-tracked. I focused on code execution in worker and handling imports and code editors, which will be added to taht prototype when I catch a bit of free time to code.

**Comment by hrgdavor** [comment]
I am tracking the discussions, and like very much how far manifold has gone already. Looking fwd when I can find some time to play with wasm. And these extra steps for performance look very interesting.

**Comment by hrgdavor** [comment]
@elalish I have starting playing with manifold and am able to run it in jscad.app. I maybe asking too soon but I would like to in the future explore using SharedArrayBuffer when getting mesh data from manifold. I have not yet come to concrete need for it, as I am still at early explorative phase, just wondering if someone has found it useful with manifold.

**Reply by elalish** [reply]
Awesome! There's some discussion of it here: https://github.com/elalish/manifold/pull/653#issuecomment-1992185723, but I'm not sure that's exactly what you mean. How would you be hoping to use it?

**Reply by hrgdavor** [reply]
One reason is getting mesh data into SharedArrayBuffer is so I can send it to main thread to render, but also have it ready if I run 3mf export inside the worker. A workaround could be simply using transferable, and returning the ArrayBuffer back from main thread after sending to gpu (if it is possible to give it away and back via transferable).

I am otherwise excited about SharedArrayBuffer and looking for ideas :) how to use them. I have some vague idea that with SharedArrayBuffer I could make a different WASM for producing some procedural shapes and have a performant way to send them to manifold wasm for further operations. Consider it more like brainstorming throw away ideas, I may be missing the target, and likely thinking about optimizations way to soon. I need to tickle my brain with grandiose ideas before going back to making something simple/boring but useful :D :D :D .

**Reply by elalish** [reply]
Love it! Anyway, I'm no ShareArrayBuffer expert, so let me know what you find and what might make a good addition to the library. I'm also thinking about adding our `worker.ts` to our npm package so people can easily get Manifold scripts evaluated. Would that be useful to you? It currently hands back GLB and 3MF, but we could talk about a different interface too.

**Reply by hrgdavor** [reply]
I need to explore a bit more, to get a better feeling of the requirements. 

I am ok with the idea of such a sandbox for playing with the library, but I want it to be just that, for initial playing. Instead indefinitely adding features to the script env runner, I would push users to convert to regular JS projects and use the ecosystem. 

on this example you can see what it takes to convert jscad V1 code to a normal JS file (imports not pruned, it is compat shim)
https://3d.hrg.hr/jscad/v1-support/#test.csg.js

and there is test branch that supports manifold.
https://3d.hrg.hr/jscad/manifold-test/#test.manifold.js
Manifoldcad env for prototyping is fine as it also handles memory, and that could be all some users need, but any that are more like me will want to get out of the sandbox and shoot them self in the foot while learning :)

I have a worker that handles complex js projects rather well, and I am close to making working with worker as simple as calling async methods `let something = await worker.doSomeModeling({...params})`

I dream big, but even as slowly as jscadui is progressing, nice things useful things are coming out of it :) ... and personally, I enjoy even more making 3d models for 3d printing using JS (I just prefer JS personally and like to share interactive stuff online).

**Reply by hrgdavor** [reply]
Also @elalish one huge thing became supported across brosers: https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/transferControlToOffscreen
- 2018-09-04 - Chrome
- 2018-09-25 - Opera
- 2020-01-15 - Edge
- 2022-09-20 - Firefox
- 2023-03-27 - Safari

**Reply by hrgdavor** [reply]
@elalish I am looking to support GLB imort(limited ofc) and export for jscad.
